### PR TITLE
Enable Lifecycle API on Prometheus 2.0.0-beta.0 and up

### PR DESCRIPTION
**Pull request by Capitrium**
The Lifecycle API is disabled by default starting with Prometheus v2.0.0-beta.0, causing the config reloader to error out:

```
ts=2017-08-15T19:12:35Z caller=main.go:214 component=volume-watcher msg="Reloading Prometheus temporarily failed." err="Received response code %!s(int=403), expected 200" next-retry=25.169469033s
```

This PR adds a flag to Prometheus to enable the Lifecycle API on v2.0.0-beta.0 and higher, allowing the config reloader to work correctly:

```
ts=2017-08-16T15:59:20Z caller=main.go:219 component=volume-watcher msg="Prometheus successfully reloaded."
```

### CODE REVIEW

**Comment by brancz** [comment]
Actually our intention was to not expose the lifecycle APIs at all, as they have a security risk, for obvious reasons. Instead the idea was to change how the prometheus-config-reloader performs reloading in the first place, as in rather than calling the reload HTTP endpoint the sidecar can just send the SIGHUP signal to the process. We're tracking this in #553. If you want to I can guide you through implementing that if you need guidance. Feel free to do that in this PR if you are interested in taking this over. Wdyt? :slightly_smiling_face:

**Comment by Capitrium** [comment]
Ah, fair enough. I ran into this while testing the latest Prometheus beta and #480 mentioned enabling the lifecycle API, but a SIGHUP works just as well. I may take a shot at that if I can find some time for it in the next couple of days.

**Comment by brancz** [comment]
No worries. Let us know if you want us to take over, no problem! :slightly_smiling_face:

**Comment by Capitrium** [comment]
@brancz I've modified the reloader to send a SIGHUP instead of an HTTP POST, seems to be working fine; not sure if there's a better way to enable signalling between containers other than using `hostPID`, although I'm sure the code to get the prometheus pid could be cleaner :stuck_out_tongue:

**Comment by mikebryant** [comment]
Isn't using `hostPID` a big escalation in how much access we're giving prometheus?

We're currently only allowing admin level things that level of access in our clusters, not user level monitoring tools.

File: contrib/prometheus-config-reloader/main.go [hunk 1]
```diff
@@ -185,17 +187,21 @@ func (w *volumeWatcher) writeConfigMapFile(filename, content string) error {
 }
 
 func (w *volumeWatcher) ReloadPrometheus() error {
-	req, err := http.NewRequest("POST", w.cfg.reloadUrl, nil)
+	cmd := exec.Command("/bin/sh", "-c", "ps cax | grep '/bin/prometheus ' | grep -v grep | grep -o '^[ ]*[0-9]*'")
```
Comment (brancz) [inline contrib/prometheus-config-reloader/main.go]: `ps` is not really magic, we can just read the procfs and find the process ourselves without relying on `ps` and `grep`.

Comment (mikebryant) [inline contrib/prometheus-config-reloader/main.go]: Also, what happens if there's more than one prometheus instance on the same host?

Comment (brancz) [inline contrib/prometheus-config-reloader/main.go]: The PID namespace is shared only within containers of the same Pod, so selecting the first process running the `/bin/prometheus` binary should be safe.

Comment (brancz) [inline contrib/prometheus-config-reloader/main.go]: Ah I just saw the hostPID part! Thanks for pointing it out. Yes that's not something we want to do, it is definitely an escalation risk.

**Comment by brancz** [comment]
Turns out this space is not as easy as it looks at first sight. I talked to the container linux team and they told me that this is a fragile piece, as for example docker and rkt handle these situation differently. Only starting Docker 1.13.1 (which is not supported by Kubernetes yet) will enable us to share PID namespace between containers of the same Pod (upstream PR regarding this: https://github.com/kubernetes/kubernetes/pull/45236).

Sorry for all the forth and back, it turns out that enabling the lifecycle API will stay being our best option for the time being. Let's go with that instead then.

**Comment by Capitrium** [comment]
No worries! I kind of figured this would end up being the case; I've reverted back to simply enabling the lifecycle api.

File: pkg/prometheus/statefulset.go [hunk 2]
```diff
@@ -327,6 +327,13 @@ func makeStatefulSetSpec(p v1alpha1.Prometheus, c *Config, ruleConfigMaps []*v1.
 			"-storage.tsdb.path=/var/prometheus/data",
 			"-storage.tsdb.retention="+p.Spec.Retention,
 		)
+
+		if len(version.Pre) > 0 && version.Pre[0].VersionStr == "alpha" {
```
Comment (brancz) [inline pkg/prometheus/statefulset.go]: we only need to support the latest beta, if I understand correctly, we can drop this as we don't support the alphas anymore, feel free to remove them from the compatibility matrix and only keep the latest beta

Comment (Capitrium) [inline pkg/prometheus/statefulset.go]: Done and done (figured this would be the case as well :simple_smile:)

**Comment by brancz** [comment]
Thanks a lot! And sorry for the forth and back!

**Pull request merged**